src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -370,14 +369,6 @@ Pipe ReadFromMergeTree::readFromPoolParallelReplicas(
             actions_settings, block_size_copy, reader_settings, virt_column_names);
 
         auto source = std::make_shared<MergeTreeSource>(std::move(processor));
-
```

Comment:
I'm not sure if removing this logic is correct in the non parallel replicas case.

In fact, I think we could use this exact part to report the rows to read also in parallel replicas with something like:
```diff
-        if (i == 0 && !client_info.collaborate_with_initiator)
+        if (i == 0 && (!client_info.collaborate_with_initiator || client_info.number_of_current_replica == 0))
             source->addTotalRowsApprox(total_rows);
```

That way we report the the total rows only on one replica (the first) and avoid the problem while keeping a good report. It'd make all other changes in the PR unnecessary (including the protocol change). WDYT?

Reply:
Nice one line fix. In this sense, I like it. But ... 

(1) it's clear that this approach doesn't cover the difference between replicas data, which can be arbitrary.  In extreme case (most probably artificial), data on replicas can entirely different, - can be achieved by stopping fetches (did it for testing purpose [here](https://github.com/ClickHouse/ClickHouse/blob/482a484a9c79be66d0f7f8d807799f222b885118/tests/integration/test_parallel_replicas_distributed_read_from_all/test.py))

(2) also, not sure how healthy will behave the progress bar when got much less total rows than actually are read. And I'd prefer to avoid getting complains about unhealthy progress bar with running parallel replicas

Reply:
IMO parallel replicas should read only and exactly the parts known by the coordinator, which in fact should probably be the one knowing how many rows are expecting to be read and not depend on any of the replicas. Is there any way that the coordinating server (and not the replicas) could report this value based on its analysis? It already knows which marks it wants to be read by the replicas

Reply:
It looks to me that you've just described the implementation in this PR. 
In short: the parallel replicas protocol was extended to send from replicas, among ranges to read, also the number of rows in ranges for each part. This happens during the announcement phase. After the coordinator receives announcements from all replicas involved in query execution, it has a set of ranges to read, and the total number of rows in them. And this number is used for reporting progress.